From 434b89e648265652cdda592894c06c734ddafa5a Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Tue, 19 Aug 2014 19:30:51 -0400
Subject: [scsi] qla4xxx: Export sysfs DDBs from DPC handler

Message-id: <1408476658-5056-14-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88346
O-Subject: [PATCH RHEL 7.1 e-stor 13/20] qla4xxx: Export sysfs DDBs from DPC handler
Bugzilla: 1089349
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Nilesh Javali <nilesh.javali@qlogic.com>

scsi commit id 906cbf3d2cef76a6e23207647cbbbc66023886f9

Issue:
modprobe qla4xxx is killed by systemd due to timeout.

Solution:
The exporting of sysfs DDBs from qla4xxx_probe_adapter added delay of
approximately 15s due to which system-udevd killed the modprobe of the
driver. Added fix to export the sysfs DDBs from the DPC handler.

Signed-off-by: Nilesh Javali <nilesh.javali@qlogic.com>
Signed-off-by: Vikas Chaudhary <vikas.chaudhary@qlogic.com>
Reviewed-by: Mike Christie <michaelc@cs.wisc.edu>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/qla4xxx/ql4_def.h b/drivers/scsi/qla4xxx/ql4_def.h
index 73a5022..8f6d0fb 100644
--- a/drivers/scsi/qla4xxx/ql4_def.h
+++ b/drivers/scsi/qla4xxx/ql4_def.h
@@ -601,6 +601,7 @@ struct scsi_qla_host {
 #define DPC_HA_NEED_QUIESCENT  22 /* 0x00400000 ISP-82xx only*/
 #define DPC_POST_IDC_ACK  23 /* 0x00800000 */
 #define DPC_RESTORE_ACB   24 /* 0x01000000 */
+#define DPC_SYSFS_DDB_EXPORT  25 /* 0x02000000 */
 
  struct Scsi_Host *host; /* pointer to host data */
  uint32_t tot_ddbs;
diff --git a/drivers/scsi/qla4xxx/ql4_glbl.h b/drivers/scsi/qla4xxx/ql4_glbl.h
index 2035475..5f58b45 100644
--- a/drivers/scsi/qla4xxx/ql4_glbl.h
+++ b/drivers/scsi/qla4xxx/ql4_glbl.h
@@ -281,6 +281,7 @@ int qla4_83xx_get_port_config(struct scsi_qla_host *ha, uint32_t *config);
 int qla4_83xx_set_port_config(struct scsi_qla_host *ha, uint32_t *config);
 int qla4_8xxx_check_init_adapter_retry(struct scsi_qla_host *ha);
 int qla4_83xx_is_detached(struct scsi_qla_host *ha);
+int qla4xxx_sysfs_ddb_export(struct scsi_qla_host *ha);
 
 extern int ql4xextended_error_logging;
 extern int ql4xdontresethba;
diff --git a/drivers/scsi/qla4xxx/ql4_os.c b/drivers/scsi/qla4xxx/ql4_os.c
index 50109d6..6446956 100644
--- a/drivers/scsi/qla4xxx/ql4_os.c
+++ b/drivers/scsi/qla4xxx/ql4_os.c
@@ -4565,6 +4565,7 @@ static void qla4xxx_timer(struct scsi_qla_host *ha)
       test_bit(DPC_LINK_CHANGED, &ha->dpc_flags) ||
       test_bit(DPC_HA_UNRECOVERABLE, &ha->dpc_flags) ||
       test_bit(DPC_HA_NEED_QUIESCENT, &ha->dpc_flags) ||
+      test_bit(DPC_SYSFS_DDB_EXPORT, &ha->dpc_flags) ||
       test_bit(DPC_AEN, &ha->dpc_flags)) {
   DEBUG2(printk("scsi%ld: %s: scheduling dpc routine"
          " - dpc flags = 0x%lx\n",
@@ -5432,6 +5433,11 @@ dpc_post_reset_ha:
     qla4xxx_relogin_all_devices(ha);
   }
  }
+ if (test_and_clear_bit(DPC_SYSFS_DDB_EXPORT, &ha->dpc_flags)) {
+  if (qla4xxx_sysfs_ddb_export(ha))
+   ql4_printk(KERN_ERR, ha, "%s: Error exporting ddb to sysfs\n",
+       __func__);
+ }
 }
 
 /**
@@ -8409,7 +8415,7 @@ exit_ddb_del:
  *
  * Export the firmware DDB for all send targets and normal targets to sysfs.
  **/
-static int qla4xxx_sysfs_ddb_export(struct scsi_qla_host *ha)
+int qla4xxx_sysfs_ddb_export(struct scsi_qla_host *ha)
 {
  struct dev_db_entry *fw_ddb_entry = NULL;
  dma_addr_t fw_ddb_entry_dma;
@@ -8847,11 +8853,8 @@ skip_retry_init:
   ql4_printk(KERN_ERR, ha,
       "%s: No iSCSI boot target configured\n", __func__);
 
- if (qla4xxx_sysfs_ddb_export(ha))
-  ql4_printk(KERN_ERR, ha,
-      "%s: Error exporting ddb to sysfs\n", __func__);
-
-  /* Perform the build ddb list and login to each */
+ set_bit(DPC_SYSFS_DDB_EXPORT, &ha->dpc_flags);
+ /* Perform the build ddb list and login to each */
  qla4xxx_build_ddb_list(ha, INIT_ADAPTER);
  iscsi_host_for_each_session(ha->host, qla4xxx_login_flash_ddb);
  qla4xxx_wait_login_resp_boot_tgt(ha);
-- 
1.7.1